packages/cache/src/lib/CacheMiddleware.ts

```diff
@@ -0,0 +1,727 @@
+      if (isRemoveByHookPayload(payload)) {
+        const { hook } = payload;
+        const { errors } = await cache[Method.Remove]({ method: Method.Remove, key, path: ['value', ...path], type, hook, errors: [] });
+
+        payload.errors = [...payload.errors, ...errors];
+      }
+
+      if (isRemoveByValuePayload(payload)) {
+        const { errors } = await cache[Method.Remove]({ method: Method.Remove, key, path: ['value', ...path], type, value, errors: [] });
+
+        payload.errors = [...payload.errors, ...errors];
+      }
+    }
+
+    return payload;
+  }
+
+  @PostProvider()
+  public async [Method.Update]<Value = StoredValue>(payload: Payloads.Update<StoredValue, Value>): Promise<Payloads.Update<StoredValue, Value>> {
+    const { key, hook } = payload;
+    const { provider: cache } = this.context;
+    const getPayload = await cache[Method.Get]({ method: Method.Get, key, path: [], errors: [] });
+
+    payload.errors = [...payload.errors, ...getPayload.errors];
+
+    if (isPayloadWithData<CacheMiddleware.Document<StoredValue>>(getPayload) && (await this.checkNotExpired(getPayload.data, key))) {
+      const newValue = await hook(getPayload.data.value, key);
+      const { errors } = await this[Method.Set]({ method: Method.Set, key, value: newValue, path: [], errors: [] });
+
+      payload.errors = [...payload.errors, ...errors];
+    }
+
+    return payload;
+  }
+
+  private async checkNotExpired(data: CacheMiddleware.Document<unknown>, key: string) {
+    if (!this.context.ttl || !this.context.ttl.enabled) return true;
+    if (new Date().getTime() - new Date(data.created).getTime() <= (this.context.ttl.timeout ?? 1000)) return true;
+
+    const { provider: cache } = this.context;
+
+    await cache[Method.Delete]({ method: Method.Delete, key, path: [], errors: [] });
+
+    return false;
+  }
+
+  private async populateCache() {
+    const { provider: cache } = this.context;
+
+    await cache[Method.Clear]({ method: Method.Clear, errors: [] });
+
+    const { data } = await this.provider[Method.Entries]({ method: Method.Entries, data: {}, errors: [] });
+
+    if (data) {
+      await this[Method.SetMany]({
+        method: Method.SetMany,
+        entries: Object.entries(data).map(([key, value]) => ({ key, value, path: [] })),
+        overwrite: true,
+        errors: []
+      });
+    }
+  }
+
+  private async fetchCache() {
+    if (this.context.fetchAll || this.context.fetchAll === undefined) {
+      await this.populateCache();
+    }
+
+    if (this.context.polling && this.context.polling.enabled) {
+      this.startPolling();
+    }
+  }
+
+  private startPolling() {
+    this.pollingInterval = setInterval(() => this.populateCache(), this.context.polling?.interval ?? 10000);
+    addExitCallback(() => {
+      clearInterval(this.pollingInterval);
+      this.pollingInterval = undefined;
+    });
+  }
+}
+
+export namespace CacheMiddleware {
+  export interface Document<StoredValue> {
```

Comment:
```suggestion
  export interface Data<StoredValue> {
```
Unless you think otherwise, I feel this is more appropriate.

Reply:
I kinda prefer document, especially as I may add more to it, such as `version`

Reply:
That's fine, it's just the first thing I think about when I hear document is mongo, lol